### PR TITLE
Feature/jdbc hometask

### PR DESCRIPTION
**Pull request by Arthur2224**
Summary:

-     Implemented Ticket as a database entity.
-     Improved relationships between Ticket and Client.
-     Enhanced Ticket ID generation.
-     Added Docker Compose configuration for hosting the database.
-     Implemented Ticket DAO pattern for better data handling.

Changes:

-     Created Ticket entity with relevant fields and relations to Client.
-     Refactored Ticket ID generation logic.
-     Added Docker Compose file for setting up PostgreSQL.
-     Added TicketDAO to manage database interactions for Ticket.

Notes:

-     The PR includes the necessary configurations for Docker to easily run the database locally.
-     The TicketDAO pattern was introduced to encapsulate database operations for Ticket entities.
-     During the database initialization process, the script also adds some initial records to both tables (clients and users).

Commands to set up DB via Docker:

-  docker-compose up --build -d

- docker ps //ensure thats container running in the right way

attels

Example of output:

attels
attels
attels
attels
attels
attels
attels

### CODE REVIEW

File: src/main/java/org/example/DAOs/DAO.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.example.DAOs;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface DAO<T> {
+    Optional<T> findById(long id);
+    List<T> getAll();
+    void save(T t);
+    void update(T t, T temp);
```
Comment (Darya-Drutsko) [inline src/main/java/org/example/DAOs/DAO.java]: why 2 params here?

File: src/main/java/org/example/DAOs/ClientDAO.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    @Override
+    public Optional<Client> findById(long id) {
+        String sql = "SELECT * FROM clients WHERE id = ?";
+        try (PreparedStatement statement = connection.prepareStatement(sql)) {
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                Client client = Client.builder()
```
Comment (Darya-Drutsko) [inline src/main/java/org/example/DAOs/ClientDAO.java]: we could extract it to mapper class

File: src/main/java/org/example/DAOs/ClientDAO.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            if (rowsAffected == 0) {
+                throw new DAOException("Failed to save client: No rows affected");
+            }
+        } catch (SQLException e) {
+            throw new DAOException("Failed to save client: " + client, e);
+        }
+    }
+
+    public void update(Client client, Client updatedClient) {
+        String sql = "UPDATE clients SET name = ?, email = ?, updated_at = ? WHERE id = ?";
```
Comment (Darya-Drutsko) [inline src/main/java/org/example/DAOs/ClientDAO.java:71]: so do we need as a param for method the whole object of previous(non updated) version?

File: src/main/java/org/example/DAOs/ClientDAO.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            if (rowsUpdated == 0) {
+                throw new DAOException("Failed to update client by ID: " + client.getId());
+            }
+        } catch (SQLException e) {
+            throw new DAOException("Failed to update client by ID: " + client.getId(), e);
+        }
+    }
+
+    public void delete(Client client) {
+        TicketDAO ticketDAO=new TicketDAO();
```
Comment (Darya-Drutsko) [inline src/main/java/org/example/DAOs/ClientDAO.java]: Add ticketDao to separate field in class

File: src/main/java/org/example/managers/TicketManager.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
+package org.example.managers;
+
+import org.example.entities.Ticket;
+import org.example.services.TicketService;
+
+import java.util.List;
+import java.util.Optional;
+
+public class TicketManager {
```
Comment (Darya-Drutsko) [inline src/main/java/org/example/managers/TicketManager.java]: what is the difference between manager and service classes?

File: src/main/java/org/example/DAOs/ClientDAO.java (hunk 3 above)
Comment (Arthur2224) [reply src/main/java/org/example/DAOs/ClientDAO.java:71]: actually no, it's enough to use only object's id of previous version

File: src/main/java/org/example/managers/TicketManager.java (hunk 5 above)
Comment (Arthur2224) [reply src/main/java/org/example/managers/TicketManager.java]: service encapsulates the business logic required for the app to perform tasks.  Meanwhile manager class is responsible for coordinating multiple services or operations to handle complex business processes. So it's not ideal to use manager as additional layer above the service class

**Review by Darya-Drutsko — APPROVED** [review]

**Pull request merged**